### ISSUE TITLE
Extracting flatten/unflatten into lightweight package?

### ISSUE THREAD

**Issue by oschulz** [invenia/ParameterHandling.jl#43]
In the spirit of creating lightweight interface-defining packages (see TuringLang/Bijectors.jl#199 which resulted in [InverseFunctions.jl](https://github.com/JuliaMath/InverseFunctions.jl) and [ChangesOfVariables.jl](https://github.com/JuliaMath/ChangesOfVariables.jl)):

While adding an interface-test utility to ChangesOfVariables.jl today, I needed a helper function [_to_realvec_and_back](https://github.com/JuliaMath/ChangesOfVariables.jl/blob/b2bd500dc7566a06d72f6be9fd4699faa1d2c708/src/test.jl#L4) - which does exactly what `ParameterHandling.flatten` does. And I needed it for exactly what #27 is requesting. :-)

It would be nice not to reinvent this, but `ParameterHandling` would of course be way to heavy a dependency for `ChangesOfVariables` - and even if it wasn't, we'd end up with a circular dependency since `Bijectors` will use `ChangesOfVariables` soon, so `ParameterHandling` [will too](TuringLang/Bijectors.jl#199).

I think the flatten/unflatten functionality of `ParameterHandling` does something very fundamental, and is orthogonal to it's variable transformation capabilities. Would you consider splitting it out into a lightweight package (basically the contents of `flatten.jl`)?

A truly lightweight recursive flatten/unflatten interface package could IMHO find use in many places in the ecosystem. `flatten` may be a bit too generic a name for the function (we probably have several `flatten`s in the ecosystem), but how about `flatten_and_back` or so?

CC @willtebbutt, @paschermayr, @devmotion

**Comment by oschulz** [comment]
> How would you feel about depending on ParameterHandling once #42 goes in?

I think even with #42 it think it would still be too heavy - the idea is to keep ChangesOfVariables super-lightweight, so we can convince packages to depend on it, and the stuff in `parameters.jl` is really a different thing - here, I think, ParameterHandling would depend on ChangesOfVariables, not the other way round. And ParameterHandling is a more opinionated package than ChangesOfVariables.

But now that I think about it, since it looks like we need flatten/unflatten functionality in ChangesOfVariables, how would you like moving that there? It's not much more than I have in the draft of JuliaMath/ChangesOfVariables.jl#2 anyway, and I think it would fit into the theme of the package - it *is* a change of variables, after all (with a logabsdet-jacobian of 0, of course).

@devmotion, what do you think?

**Comment by willtebbutt** [comment]
> so one might want to have a way to mark parameters as active/constant.

We actually already have a way to mark parameters as being constant (see `fixed`), although doing things the other way around (marking things as "active") is something that we don't currently have support for -- although we could almost certainly achieve using a trait-based system.

**Comment by oschulz** [comment]
@devmotion pointed me to https://github.com/JuliaDiff/FiniteDifferences.jl/blob/main/src/to_vec.jl

There's probably equivalents of this in several places in the ecosystem, internally. If we had a nice, lightweight central API for this, people could make their types support it, that would make a lot of AD a bit easier. It's handling integers and constant parameters though that makes it less straightforward than I thought initialy - but I think it could (and maybe should) still be decoupled from actual value transformations.

**Comment by ToucheSir** [comment]
For the interested, here's how Optimisers.jl handles reverse mode friendly (un)flattening: https://github.com/FluxML/Optimisers.jl/blob/master/src/destructure.jl. We opted for the conservative approach and only consider non-integer numeric arrays (due in part to Flux's design constraints). Something like `fixed` would be interesting, but the main challenge with such a wrapper is making it transparent to non-parameter handling code.

**Comment by oschulz** [comment]
> For the interested, here's how Optimisers.jl handles reverse mode friendly (un)flattening

This is build on top of Functors.jl mainly, right?

**Comment by oschulz** [comment]
While revamping ForwardDiffPullbacks, I had to add "invent" [yet another flatten/unflatten mechanism](https://github.com/oschulz/ForwardDiffPullbacks.jl/blob/main/src/flatten.jl). To get full performance and type stability (this needs to be really fast and allocation-free with deeply nested structures, flattening to static vectors), I ended up with

* `flatten(x)`
* `unflatten(x_orig, x_flat)`
* `unflatten_tangent(x_orig, dx_flat)`

An approach like `x_flat, reconstruct_function = flatten(x)` didn't work out performance/type-stability wise, there was trouble when things were nested more deeply (though maybe I did it wrong, I also tried via [Flatten.jl](https://github.com/rafaqz/Flatten.jl) but that also didn't work out). Also the `x_flat, re = ...` pattern didn't provide tangent-unflatten capability.

The problem is that with these flatten/unflatten capabilities "hidden" in several packages in different ways, it's near impossible for users to specify flatten/unflatten for types that need special handling without creating a dependency nightmare. I wonder if we could come up with a generic lightweight thing similar to ChainRulesCore that would satisfy all use cases?

**Comment by paschermayr** [comment]
> An approach like x_flat, reconstruct_function = flatten(x) didn't work out performance/type-stability wise

I guess you could make methods in case `reconstruct_function` is not needed.

**Comment by oschulz** [comment]
I think I should link to the current discussion regarding `ConstructionBase.getfields`/`ConstructionBase.getproperties` (JuliaObjects/ConstructionBase.jl#54) here, I feel these issues are very connected.

**Comment by ToucheSir** [comment]
> An approach like `x_flat, reconstruct_function = flatten(x)` didn't work out performance/type-stability wise, there was trouble when things were nested more deeply (though maybe I did it wrong, I also tried via [Flatten.jl](https://github.com/rafaqz/Flatten.jl) but that also didn't work out). Also the `x_flat, re = ...` pattern didn't provide tangent-unflatten capability.

The implementation isn't type stable because of unrelated factors (mostly caching in Functors.jl), but [`Optimisers.destructure`](https://github.com/FluxML/Optimisers.jl/blob/master/src/destructure.jl) does basically this. Saving some auxiliary state during flattening pays off when # reconstructions > # of flattenings. What does _not_ work well is returning a plain closure, but I'm preaching to the choir here :)

> I wonder if we could come up with a generic lightweight thing similar to ChainRulesCore that would satisfy all use cases?

I've thought about this and it's just genuinely hard because of how broad and nuanced "(un)flattening" is. Some examples to chew on:
- Flux wants to flatten arbitrary structs with non-flattenable fields without forcing them to be generic in those fields like Flatten.jl requests.
- Accounting for parameters that are "shared" or "tied" is hard. Annotating them with a wrapper in the source parameter collection can work, but messes with dispatch in unsuspecting downstream consumers of that collection. Trying to do so automatically turned out to be a rabbit hole we didn't want to mess with in Optimisers.jl.
- The same point about wrappers and dispatch applies to constrained parameters as exist in ParameterHanding.jl or ModelParameters.jl.
- It may not be possible to derive the fixed-ness of a given subtree's children via the type alone. Here too we see a range of very distinct solutions, including `ParameterHanding.fixed` and `Optimisers.trainable`. The latter brings up another point: for many applications we can not assume that all parameter-visiting traversals of an object tree(/DAG) will care about the same set of fields flattening does. This is the motivation behind [StructWalk.jl](https://github.com/chengchingwen/StructWalk.jl)'s `WalkStyle`.

**Comment by oschulz** [comment]
I agree, it's a tricky, multi-faceted problem. But I still think we have too many competing solutions in the ecosystem right now. Maybe one could somehow factorize this into an API for struct-developers that allows them to "annotate" their structs so we need less heuristics/guesswork, and a set of flatten/unflatten APIs for engine/algorithms-developers that make use of those? I don't have a concrete proposal, I just feel that some ChainRulesCore-like (in spirit, not functionality) standard in this area is really missing in the ecosystem (or possibly the language itself) at the moment.

**Comment by ToucheSir** [comment]
AFAIK something like https://github.com/rafaqz/FieldMetadata.jl could be that standard, but the note at the top of the README seems to suggest differently :/

**Comment by oschulz** [comment]
@rafaqz can we pull you in here as well?

**Comment by rafaqz** [comment]
FieldMetadata.jl is a cool idea, but it sets metadata by defining new functions on an object. So you have method table state to think about if you ever want to change anything during use. Its also a lot of confusing and fragile macros for people to understand, that dont scale to organisation/research group level use very well.

ModelParameters.jl is a better solution. You can do most of the same things but state is contained in the object, and has a Tables.jl interface.

What do you need from this package that you can't do with ModelParameters?

**Comment by rafaqz** [comment]
Additionally, truly lightweight recursive rebuilding is possible with my [PR to Accessors.jl](https://github.com/JuliaObjects/Accessors.jl/pull/23). And thats probably the most generic place to pit it. It hasn't been merged because Im too busy and we can't get it type stable because of issues with Base no longer inferring recursive methods.

**Comment by oschulz** [comment]
> What do you need from this package that you can't do with ModelParameters?

If you have a type that needs a bit of special handling, and you want it to be compatible with a wide variety of Julia ML, statistics & friends packages, and you want GPU + GPU support, you currently have to depend on and specialize functionality defined in: Adapt, ConstructionBase, Functors, ParameterHandling, and possibly a few others. To my knowledge, none of those draw on each other for their default implementations. This doesn't compose well and forces packages to either take on a lot of dependencies or not be compatible with parts of the ecosystem. And it can result in a lot of boilerplate code.

As a type developer, I'll probably just want to implement a simple, closure-free API like

```julia
get_raw_contents(x::MyType)::Union{Real,Tuple,AbstractArray,NamedTuple}
get_semantic_contents(x::MyType)::Union{Real,Tuple,AbstractArray,NamedTuple}
reconstruct_from(::Type{<:MyType}, ::ResultOfGetRawContents)
reconstruct_from(::Type{<:MyType}, ::ResultOfGetSemanticContents)
```

The contract here would be

```julia
reconstruct_from(typeof(x), get_raw_contents(x) == x
reconstruct_from(typeof(x), get_semantic_contents(x) == x
```

`reconstruct_from` should accept contents with a different numeric precision and array types as the original `x`, of course, if supported by the type.

A package like ConstructionBase would seem a natural place to host such an API.

`get_raw_contents`  and `get_semantic_contents` could return tuples, `NamedTuple`s, `Real`s (even all `Number`s?) and `Array`s as-is, and return a `NamedTuple` for structs - `get_raw_contents` could use `fieldcount`/`fieldname/`getfield` and `get_semantic_contents` could use `propertynames`/`getproperty`.

A package like Adapt would probably want to use `get_raw_contents` for it's default implementation, whereas packages like Functors and ParameterHandling would want `get_semantic_contents` I guess.

**Comment by rafaqz** [comment]
Yes ConstructionBase.jl is the natural place for this, and sharing these base methods across more packages was the original reason for it to be written. Flatten.jl, Setfield.jl and I think BangBang.jl all needed it. Fixing any inadequacies it has to make it useful in these other packages is surely within scope.

In your schema I guess `getrawcontents` == `ConstuctionsBase.getfields` (in the current PR by @jw3126) and `get_semantic_contents` == `ConstructionBase.getproperties` ?

Also, Adapt.jl is already kind of redundant. You can replace it with Flatten.jl or Accessors.jl with that PR. I do this in some GPU based packages like DynamicGrids.jl, but the current problem with types stability in base makes this less of an option than it used to be.

**Comment by jw3126** [comment]
> how would you foresee handling multiple different versions of get_semantic_contents for the same type?

IMO multiple `get_semantic_contents` on the same type are not the job of ConstructionBase. Analogous to Base not providing multiple variants of `getproperty` on a single type. Instead I think lenses are a great abstraction for this, so you could use Accessors.jl or Setfield.jl

**Comment by oschulz** [comment]
> @oschulz the closure-free API proposal above looks very interesting, one question for now: how would you foresee handling multiple different versions of get_semantic_contents for the same type?

Thanks!

Multiple different versions of get_semantic_contents for the same type in what respect? Could you give a quick example?

**Comment by ToucheSir** [comment]
I put a relevant example in the comment above, but the gist is that not all traversals we make over a nested object ~~DAG~~ tree (I'm being cheeky here, but handling "shared" nodes is a big part of the Functors codebase) will want to access the same set of fields. Thus defining a single `get_semantic_contents` function has limited value for our use cases unless the definition of that function is sufficiently general. But then if it's too general, there doesn't seem to be much of a difference between `get_semantic_contents` and `get_raw_contents`, AIUI.

This is why I mentioned StructWalk earlier: by parameterizing the (un)flatten function with the type of traversal being performed (`WalkStyle`), the traversal code can be decoupled from the code that extracts the contents of each node. Generic fallbacks mean that no convenience is lost for types that return the same contents regardless of the purpose of the traversal.

**Comment by oschulz** [comment]
>  but handling "shared" nodes is a big part of the Functors codebase

That would be part of the nested walk scheme relevant to the application though, right, and orthogonal to the "get content of this type" API?

> by parameterizing the (un)flatten function with the type of traversal being performed

Ah, I think I get it @ToucheSir . So you mean instead of having `get_raw_contents(x)` and `get_semantic_contents(x)` we'll need something like `get_contents(x, decomposition_style)`, with a matching `reconstruct_from(T, content, decomposition_style)`?

Do you think there would be a finite number of such "decomposition styles" and/or a kind of hierarchy between them so a type developer won't need to know (and specialize for) all of them?

**Comment by ToucheSir** [comment]
> Do you think there would be a finite number of such "decomposition styles" and/or a kind of hierarchy between them so a type developer won't need to know (and specialize for) all of them?

Yes. The default could be identical to `get_semantic_contents(x)`. I'll pick up the rest on the ConstructionBase thread.
